instrumentation/opentelemetry-instrumentation-aws-lambda/src/opentelemetry/instrumentation/aws_lambda/__init__.py

```diff
@@ -110,223 +110,694 @@ def custom_event_context_extractor(lambda_event):
-        logger.debug(
-            "Extracting context from Lambda Event failed: either enable X-Ray active tracing or configure API Gateway to trigger this Lambda function as a pure proxy. Otherwise, generated spans will have an invalid (empty) parent context."
-        )
-    if not isinstance(headers, dict):
-        headers = {}
-    return get_global_textmap().extract(headers)
 
+    return lambda timeout: logger.warning(
+        "%s was missing `force_flush` method. "
+        "This is necessary in case of a Lambda freeze and would exist in the "
+        "OTel SDK implementation.",
+        provider_type,
+    )
 
-def _determine_parent_context(
-    lambda_event: Any,
-    event_context_extractor: Callable[[Any], Context],
-    disable_aws_context_propagation: bool = False,
-) -> Context:
-    """Determine the parent context for the current Lambda invocation.
 
-    See more:
-    https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/instrumentation/aws-lambda.md#determining-the-parent-of-a-span
+def determine_tracer_provider(
+    tracer_provider_override: Optional[Any],
+) -> FlushableTracerProvider:
+    """
+    Determine the TracerProvider to use for the instrumentation.
+
+    The TracerProvider must have a `force_flush` method to be used in the instrumentation.
+    If the TracerProvider does not have a `force_flush` method,
+    a dummy `force_flush` method is added that logs a warning.
 
     Args:
-        lambda_event: user-defined, so it could be anything, but this
-            method counts it being a map with a 'headers' key
-        event_context_extractor: a method which takes the Lambda
-            Event as input and extracts an OTel Context from it. By default,
-            the context is extracted from the HTTP headers of an API Gateway
-            request.
-        disable_aws_context_propagation: By default, this instrumentation
-            will try to read the context from the `_X_AMZN_TRACE_ID` environment
-            variable set by Lambda, set this to `True` to disable this behavior.
-    Returns:
-        A Context with configuration found in the carrier.
+        tracer_provider_override: Optional TracerProvider to use for the instrumentation.
+
+    Returns: TracerProvider that is flushable.
     """
-    parent_context = None
 
-    if not disable_aws_context_propagation:
-        xray_env_var = os.environ.get(_X_AMZN_TRACE_ID)
+    if not isinstance(tracer_provider_override, TracerProvider):
+        tracer_provider = get_tracer_provider()
+    else:
+        tracer_provider = tracer_provider_override
 
-        if xray_env_var:
-            parent_context = AwsXRayPropagator().extract(
-                {TRACE_HEADER_KEY: xray_env_var}
-            )
+    if not hasattr(tracer_provider, "force_flush"):
+        tracer_provider.force_flush = provider_warning(
+            provider_type="TracerProvider"
+        )
```

Comment:
Functionally the Flushable Providers don't serve any purpose other than fixing up some type-hinting issues. 
I'll have a look back through and see if it can be optimised.